File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFilePermissionModule.java
package versioned.host.exp.exponent.modules.universal;

import android.content.Context;

import java.io.File;
import java.io.IOException;
import java.util.EnumSet;

import org.unimodules.core.ModuleRegistry;
import expo.modules.filesystem.FilePermissionModule;
import org.unimodules.interfaces.filesystem.Permission;

import expo.modules.interfaces.constants.ConstantsInterface;
import host.exp.exponent.utils.ScopedContext;

public class ScopedFilePermissionModule extends FilePermissionModule {
  private ScopedContext mScopedContext;
  private ModuleRegistry mModuleRegistry;

  public ScopedFilePermissionModule(ScopedContext scopedContext) {
    mScopedContext = scopedContext;
  }

  @Override
  protected EnumSet<Permission> getExternalPathPermissions(String path) {
    try {
      // In scoped context we do not allow access to Expo Go's directory,
      // however accessing other directories is ok as far as we're concerned.
      Context context = mScopedContext.getContext();
      String dataDirCanonicalPath = new File(context.getApplicationInfo().dataDir).getCanonicalPath();
      String canonicalPath = new File(path).getCanonicalPath();
      boolean isInDataDir = canonicalPath.startsWith(dataDirCanonicalPath + "/");
      isInDataDir |= canonicalPath.equals(dataDirCanonicalPath);
      if (shouldForbidAccessToDataDirectory() && isInDataDir) {
        return EnumSet.noneOf(Permission.class);
      }
    } catch (IOException e) {
      // Something's not right, let's be cautious.
      return EnumSet.noneOf(Permission.class);
    }
    return super.getExternalPathPermissions(path);
  }

  private boolean shouldForbidAccessToDataDirectory() {
    ConstantsInterface constantsModule = mModuleRegistry.getModule(ConstantsInterface.class);
    // If there's no constants module, or app ownership isn't "expo", we're not in Expo Go.
    return constantsModule != null && "expo".equals(constantsModule.getAppOwnership());
  }

  @Override
  public void onCreate(ModuleRegistry moduleRegistry) {
    mModuleRegistry = moduleRegistry;
  }
}

File: apps/expo-go/android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFilePermissionModule.kt
package versioned.host.exp.exponent.modules.universal

import android.content.Context
import expo.modules.kotlin.services.FilePermissionService
import host.exp.exponent.utils.ScopedContext
import java.io.File
import java.io.IOException
import java.util.EnumSet

class ScopedFilePermissionService(private val scopedContext: ScopedContext) :
  FilePermissionService() {
  override fun getPathPermissions(context: Context, path: String): EnumSet<Permission> {
    return try {
      val canonicalPath = File(path).canonicalPath
      val scopedDirectories = listOf(
        scopedContext.filesDir,
        scopedContext.cacheDir,
        scopedContext.noBackupFilesDir
      ).map { it.canonicalPath }

      if (scopedDirectories.any { canonicalPath.isInDirectory(it) }) {
        return EnumSet.of(Permission.READ, Permission.WRITE)
      }

      val dataDirCanonicalPath = File(scopedContext.context.applicationInfo.dataDir).canonicalPath
      if (canonicalPath.isInDirectory(dataDirCanonicalPath)) {
        return EnumSet.noneOf(Permission::class.java)
      }

      getExternalPathPermissions(path)
    } catch (e: IOException) {
      // Something's not right, let's be cautious.
      EnumSet.noneOf(Permission::class.java)
    }
  }

  override fun getExternalPathPermissions(path: String): EnumSet<Permission> {
    try {
      // In scoped context we do not allow access to Expo Go's directory,
      // however accessing other directories is ok as far as we're concerned.
      // Scoped context is only an Expo Go concept, so we should forbid access to data directory.
      val context = scopedContext.context
      val dataDirCanonicalPath = File(context.applicationInfo.dataDir).canonicalPath
      val canonicalPath = File(path).canonicalPath
      val isInDataDir =
        canonicalPath.startsWith("$dataDirCanonicalPath/") || (canonicalPath == dataDirCanonicalPath)
      if (isInDataDir) {
        return EnumSet.noneOf(Permission::class.java)
      }
    } catch (e: IOException) {
      // Something's not right, let's be cautious.
      return EnumSet.noneOf(Permission::class.java)
    }
    return super.getExternalPathPermissions(path)
  }

  private fun String.isInDirectory(directory: String): Boolean =
    startsWith("$directory/") || this == directory
}
